File: core/src/org/tetawex/ecf/util/PreferencesProvider.java
package org.tetawex.ecf.util;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Preferences;
import com.badlogic.gdx.utils.Json;
import org.tetawex.ecf.model.ECFPreferences;
import org.tetawex.ecf.model.LevelCompletionState;
import org.tetawex.ecf.model.Score;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Tetawex on 03.06.2017.
 */
public class PreferencesProvider {
    private static final String PREFERENCES_NAME = "ECF Preferences";
    private static final String PREFERENCES_CODE = "serialized";
    public static final int LEVELS_COUNT = 18;
    public static final int MOT_LEVELS_COUNT = 2;
    private static ECFPreferences preferences;

    public static ECFPreferences getPreferences() {
        if (preferences == null) {
            Json json = new Json();
            Preferences prefs = Gdx.app.getPreferences(PREFERENCES_NAME);
            preferences = json.fromJson(ECFPreferences.class,
                    prefs.getString(PREFERENCES_CODE));
            if (preferences == null) {
                preferences = new ECFPreferences();
                initEmptyPreferences(preferences);
            }
            int size = preferences.getLevelCompletionStateList("").size();
            if (size < LEVELS_COUNT) {
                for (int i = LEVELS_COUNT - size - 1; i <= LEVELS_COUNT; i++) {
                    preferences.getLevelCompletionStateList("").add(new LevelCompletionState(0, false, false));
                }
                preferences.getLevelCompletionStateList(("")).get(0).setUnlocked(true);
            }
            size = preferences.getLevelCompletionStateList("mot").size();
            if (size < MOT_LEVELS_COUNT) {
                for (int i = MOT_LEVELS_COUNT - size - 1; i <= MOT_LEVELS_COUNT; i++) {
                    preferences.getLevelCompletionStateList("mot").add(new LevelCompletionState(0, false, false));
                }
                preferences.getLevelCompletionStateList(("mot")).get(0).setUnlocked(true);
            }
        }
        return preferences;

    }

    public static void flushPreferences() {
        Json json = new Json();
        Preferences prefs = Gdx.app.getPreferences(PREFERENCES_NAME);
        prefs.putString(PREFERENCES_CODE, json.toJson(getPreferences()));
        prefs.flush();
    }

    private static void initEmptyPreferences(ECFPreferences preferences) {
        preferences.setSelectedLanguageCode("default");
        preferences.setSelectedCountryCode("");
        List<Score> list = new ArrayList<Score>();
        list.add(new Score(9000, "Cirno", "1-9, Icicles"));
        list.add(new Score(8000, "Blazing Borkus", "1-8, Elementalism"));
        list.add(new Score(6000, "Vivid Turtle", "1-4, The Grand Tortoise"));
        list.add(new Score(4000, "Pro WOTE Player", "1-10, Elemental Rampage"));
        list.add(new Score(3400, "Casus Beli", "1-64, Decimal Fail"));
        list.add(new Score(2007, "Blazer Hero", "1-2, Time Machine"));
        list.add(new Score(1000, "Rumia", "1-5, Decimal System"));
        list.add(new Score(900, "Cirno", "1-15, The Reckoning of Ice"));
        list.add(new Score(300, "Newbie", "1-0, Eh? Is that a level?"));
        list.add(new Score(100, "Bitter Butter", "1-64, Decimal Fail"));

        List<LevelCompletionState> levelList = new ArrayList<LevelCompletionState>();
        levelList.add(new LevelCompletionState(0, false, true));
        for (int i = 1; i < LEVELS_COUNT; i++) {
            levelList.add(new LevelCompletionState(0, false, false));
        }
        preferences.setLevelCompletionStateList(levelList, "");
        preferences.setScores(list);

        //init dlc levels
        List<LevelCompletionState> levelList2 = new ArrayList<LevelCompletionState>();
        levelList2.add(new LevelCompletionState(0, false, true));
        for (int i = 1; i < MOT_LEVELS_COUNT; i++) {
            levelList2.add(new LevelCompletionState(0, false, false));
        }
        preferences.setLevelCompletionStateList(levelList2, "mot");
    }
}

File: core/src/org/tetawex/ecf/util/PreferencesProvider.kt
package org.tetawex.ecf.util

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.utils.Json
import org.tetawex.ecf.model.ECFPreferences
import org.tetawex.ecf.model.LevelCompletionState
import org.tetawex.ecf.model.Score

import java.util.ArrayList

/**
 * Created by Tetawex on 03.06.2017.
 */
object PreferencesProvider {
    private val PREFERENCES_NAME = "ECF Preferences"
    private val PREFERENCES_CODE = "serialized"
    val LEVELS_COUNT = 18
    val MOT_LEVELS_COUNT = 15
    private var preferences: ECFPreferences? = null
    const val CURRENT_VERSION_CODE = 10

    fun getPreferences(): ECFPreferences {
        if (preferences == null) {
            val json = Json()
            val prefs = Gdx.app.getPreferences(PREFERENCES_NAME)
            preferences = json.fromJson(ECFPreferences::class.java,
                    prefs.getString(PREFERENCES_CODE))
            if (preferences == null) {
                preferences = ECFPreferences()
                initEmptyPreferences(preferences!!)
            }

            if (preferences?.version ?: 0 < 10) {
                preferences!!.scores = getDefaultScores()
            }

            if (preferences?.version ?: 0 < CURRENT_VERSION_CODE) {
                preferences!!.version = CURRENT_VERSION_CODE
            }

            var size = preferences!!.getLevelCompletionStateList("")!!.size
            if (size < LEVELS_COUNT) {
                for (i in 0 until MOT_LEVELS_COUNT - size) {
                    preferences!!.getLevelCompletionStateList("")!!.add(LevelCompletionState(0, false, false))
                }
                preferences!!.getLevelCompletionStateList("")!![0].unlocked = true
            }
            size = preferences!!.getLevelCompletionStateList("mot")!!.size
            if (size < MOT_LEVELS_COUNT) {
                for (i in 0 until MOT_LEVELS_COUNT - size) {
                    preferences!!.getLevelCompletionStateList("mot")!!.add(LevelCompletionState(0, false, false))
                }
                preferences!!.getLevelCompletionStateList("mot")!![0].unlocked = true
            }
        }
        return preferences!!
    }

    fun flushPreferences() {
        val json = Json()
        val prefs = Gdx.app.getPreferences(PREFERENCES_NAME)
        prefs.putString(PREFERENCES_CODE, json.toJson(getPreferences()))
        prefs.flush()
    }

    private fun initEmptyPreferences(preferences: ECFPreferences) {
        preferences.selectedLanguageCode = "default"
        preferences.selectedCountryCode = ""
        val list = getDefaultScores()

        val levelList = ArrayList<LevelCompletionState>()
        levelList.add(LevelCompletionState(0, false, true))
        for (i in 1 until LEVELS_COUNT) {
            levelList.add(LevelCompletionState(3, false, false))
        }
        preferences.setLevelCompletionStateList(levelList, "")
        preferences.scores = list

        //init dlc levels
        val levelList2 = ArrayList<LevelCompletionState>()
        levelList2.add(LevelCompletionState(0, false, true))
        for (i in 1 until MOT_LEVELS_COUNT) {
            levelList2.add(LevelCompletionState(0, false, false))
        }
        preferences.setLevelCompletionStateList(levelList2, "mot")
    }

    private fun getDefaultScores(): MutableList<Score> {
        val list = ArrayList<Score>()
        list.add(Score(0, "---", ""))
        list.add(Score(0, "---", ""))
        list.add(Score(0, "---", ""))
        list.add(Score(0, "---", ""))
        list.add(Score(0, "---", ""))
        list.add(Score(0, "---", ""))
        list.add(Score(0, "---", ""))
        list.add(Score(0, "---", ""))
        list.add(Score(0, "---", ""))
        list.add(Score(0, "Invisible Entry", "???"))
        return list
    }

    fun getLevelCountForCode(code: String): Int {
        return if ("mot" == code)
            MOT_LEVELS_COUNT
        else
            LEVELS_COUNT
    }
}
